waspc/src/Wasp/AppSpec/App/Auth/AuthMethods.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Wasp.AppSpec.App.Auth.AuthMethods
+  ( generateAuthMethods,
+    userSignupFieldsForEmailAuth,
+    userSignupFieldsForUsernameAuth,
+    userSignupFieldsForExternalAuth,
+    isAuthMethodExternal,
+    AuthMethod (..),
+    UsernameAndPasswordConfig (..),
+    ExternalAuthConfig (..),
+    EmailAuthConfig (..),
+  )
+where
+
+import Data.Data (Data)
+import Language.Haskell.TH
+import Language.Haskell.TH.Syntax (VarBangType)
+import Wasp.AppSpec.App.Auth.EmailVerification (EmailVerificationConfig)
+import Wasp.AppSpec.App.Auth.PasswordReset (PasswordResetConfig)
+import Wasp.AppSpec.App.EmailSender (EmailFromField)
+import Wasp.AppSpec.ExtImport (ExtImport)
+import Wasp.Util (toLowerFirst)
+
+data AuthMethod = UsernameAndPassword | Email | Google | Keycloak | GitHub deriving (Show, Enum, Bounded)
+
+data UsernameAndPasswordConfig = UsernameAndPasswordConfig
+  { userSignupFields :: Maybe ExtImport
+  }
+  deriving (Show, Eq, Data)
+
+data ExternalAuthConfig = ExternalAuthConfig
+  { configFn :: Maybe ExtImport,
+    userSignupFields :: Maybe ExtImport
+  }
+  deriving (Show, Eq, Data)
+
+data EmailAuthConfig = EmailAuthConfig
+  { userSignupFields :: Maybe ExtImport,
+    fromField :: EmailFromField,
+    emailVerification :: EmailVerificationConfig,
+    passwordReset :: PasswordResetConfig
+  }
+  deriving (Show, Eq, Data)
+
+configType :: AuthMethod -> Name
+configType UsernameAndPassword = ''UsernameAndPasswordConfig
+configType Email = ''EmailAuthConfig
+configType Google = ''ExternalAuthConfig
+configType Keycloak = ''ExternalAuthConfig
+configType GitHub = ''ExternalAuthConfig
+
+-- Generate the AuthMethods data type
+-- data AuthMethods = AuthMethods
```

Comment:
Seems to me like you ended up defining this relationship anyway above! Becauwe you defined it via `configType` at the end, only to then generate AuthMethods from that. So did you really gain anything?